tools/deploy/planner.go

```diff
@@ -0,0 +1,263 @@
+		plan.Steps = append(plan.Steps, step)
+		total -= len(step.Targets)
+	}
+	return plan
+}
+
+func (p deploymentPlanner) generateStep(
+	instances map[string]*instanceGroup,
+	maxStepSize int,
+) deploymentStep {
+	// NB(xichen): we always choose instances that are currently in the follower state first,
+	// unless there are no more follower instances, in which case we'll deploy the leader instances.
+	// This is to reduce the overall deployment time due to reduced number of leader promotions and
+	// as such we are less likely to need to wait for the follower instances to be ready to take over
+	// the leader role.
+	step := p.generateStepFromTargetType(instances, maxStepSize, followerTarget)
+
+	// If we have found some follower instances to deploy, we don't attempt to deploy leader
+	// instances in the same step even if we have not reached the max step size to avoid delaying
+	// deploying to the followers due to deploying leader instances.
+	if len(step.Targets) > 0 {
+		return step
+	}
+
+	// If we have not found any followers, we proceed to deploy leader instances.
+	return p.generateStepFromTargetType(instances, maxStepSize, leaderTarget)
+}
+
+func (p deploymentPlanner) generateStepFromTargetType(
+	instances map[string]*instanceGroup,
+	maxStepSize int,
+	targetType targetType,
+) deploymentStep {
+	step := deploymentStep{Targets: make([]deploymentTarget, 0, maxStepSize)}
+	for shardSetID, group := range instances {
+		if len(group.ToDeploy) == 0 {
+			delete(instances, shardSetID)
+			continue
+		}
+		for i, instance := range group.ToDeploy {
+			if !matchTargetType(instance.PlacementID, group.LeaderID, targetType) {
+				continue
+			}
+			target := deploymentTarget{
+				Instance:  instance,
+				Validator: p.validatorFactory.ValidatorFor(instance, group, targetType),
+			}
+			step.Targets = append(step.Targets, target)
+			group.removeInstanceToDeploy(i)
+			if maxStepSize != 0 && len(step.Targets) >= maxStepSize {
+				return step
+			}
+			break
+		}
+	}
+	return step
+}
+
+func (p deploymentPlanner) groupInstancesByShardSetID(
+	toDeploy, all instanceMetadatas,
+) (map[string]*instanceGroup, error) {
+	grouped := make(map[string]*instanceGroup, len(toDeploy))
+
+	// Group the instances to be deployed by shard set id.
+	for _, instance := range toDeploy {
+		shardSetID := instance.ShardSetID
+		group, exists := grouped[shardSetID]
+		if !exists {
+			group = &instanceGroup{
+				ToDeploy: make(instanceMetadatas, 0, 2),
+				All:      make(instanceMetadatas, 0, 2),
+			}
+		}
+		group.ToDeploy = append(group.ToDeploy, instance)
+		grouped[shardSetID] = group
+	}
+
+	// Determine the full set of instances in each group.
+	for _, instance := range all {
+		shardSetID := instance.ShardSetID
+		group, exists := grouped[shardSetID]
+		if !exists {
+			continue
+		}
+		group.All = append(group.All, instance)
+	}
+
+	// Determine the leader of each group.
+	var (
+		wg    sync.WaitGroup
+		errCh = make(chan error, 1)
```

Comment:
Sure I can record all errors. Probably can just use an error channel of size `len(group)` and process the errors at the end instead of introducing a locked multi error thing.